server/modules/scheduling/index.js

```diff
@@ -0,0 +1,181 @@
+
+  let schedule = parseResult.schedule;
+
+  if (schedule.skip)
+    return skip(skipReasons.explicitNoSchedule, source);
+
+  let expectedState = expectedStateFromSchedule(schedule, dateTime);
+
+  if (currentState === expectedState)
+    return skip(skipReasons.stateIsCorrect, source);
+
+  if (expectedState === states.on)
+    return takeAction(switchOnAction(instance), source);
+
+  return takeAction(switchOffAction(instance), source);
+
+}
+
+function switchOnAction(instance) {
+  if (instance.AutoScalingGroup) {
+    let lifeCycleStatus = getInstanceLifeCycleState(instance);
+    if (lifeCycleStatus === lifeCycleStates.outOfService)
+      return actions.putInService;
+  }
+  return actions.switchOn;
+}
+
+function switchOffAction(instance) {
+  if (instance.AutoScalingGroup) {
+    let lifeCycleStatus = getInstanceLifeCycleState(instance);
+    if (lifeCycleStatus === lifeCycleStates.inService)
+      return actions.putOutOfService;
+  }
+  return actions.switchOff;
+}
+
+function getInstanceLifeCycleState(instance) {
+  let asgInstanceEntry = _.first(instance.AutoScalingGroup.Instances.filter(i => i.InstanceId.toLowerCase() == instance.InstanceId.toLowerCase()));
+  
+  if (asgInstanceEntry) {
+    if (asgInstanceEntry.LifecycleState === 'Standby') return lifeCycleStates.outOfService;
+    if (asgInstanceEntry.LifecycleState === 'InService') return lifeCycleStates.inService;
+  }
+  
+  return lifeCycleStates.transitioning;
+}
+
+function getScheduleForInstance(instance) {
+
+  let instanceSchedule = getTagValue(instance, 'schedule');
+  if (instanceSchedule) return { parseResult: parseSchedule(instanceSchedule), source: sources.instance };
+
+  if (instance.AutoScalingGroup) {
+    let asgSchedule = getTagValue(instance.AutoScalingGroup, 'schedule');
+    if (asgSchedule) return { parseResult: parseSchedule(asgSchedule), source: sources.asg };
+  }
+  
+  return { parseResult: parseEnvironmentSchedule(instance.Environment), source: sources.environment };
+
+}
+
+function parseEnvironmentSchedule(environmentSchedule) {
+
+  if (environmentSchedule.ManualScheduleUp === false && environmentSchedule.ScheduleAutomatically === false)
+    return { success: true, schedule: { permanent: 'off' } };
+  
+  if (!(environmentSchedule.ManualScheduleUp !== true && environmentSchedule.ScheduleAutomatically === true))
+    return { success: true, schedule: { permanent: 'on' } };
+  
+  return parseSchedule(environmentSchedule.DefaultSchedule);
+
+}
+
+function expectedStateFromSchedule(schedule, dateTime) {
+  if (schedule.permanent)
+    return schedule.permanent;
+
+  let scheduleStates = schedule.map(schedule => {
+    return {
+      dateTime: later.schedule(schedule.recurrence).prev(1, dateTime),
+      state: schedule.state
+    };
+  });
+
+  let latest = _.maxBy(scheduleStates, scheduleState => scheduleState.dateTime);
+
+  return latest.state;
+}
+
+function getTagValue(instance, tagName) {
```

Comment:
Might be worth using Instance model and "getTag()" method (from taggable)
https://github.com/trainline/environment-manager/blob/master/server/models/Instance.js#L28